src/java/org/apache/cassandra/replication/CoordinatedTransfer.java

```diff
@@ -0,0 +1,431 @@
+                responses.remove(msg.from());
+                if (responses.isEmpty())
+                    trySuccess(null);
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailure failure)
+            {
+                logger.debug("{} Got failure {} from {}", logPrefix(), failure, from);
+                tryFailure(null);
+            }
+        }
+
+        DryRun dryRun = new DryRun();
+        for (InetAddressAndPort peer : peers)
+        {
+            TransferActivation activation = new TransferActivation(this, peer, true);
+            Message<TransferActivation> msg = Message.out(Verb.TRACKED_TRANSFER_ACTIVATE_REQ, activation);
+            logger.debug("{} Sending {} to peer {}", logPrefix(), activation, peer);
+            MessagingService.instance().sendWithCallback(msg, peer, dryRun);
+            SingleTransferResult result = CoordinatedTransfer.this.streams.get(msg.from());
+            if (result != null)
+                result.sentActivation();
+        }
+        dryRun.awaitUninterruptibly();
+        logger.debug("{} Dry run complete for {}", logPrefix(), peers);
+
+        // Acknowledgement of activation is equivalent to a remote write acknowledgement. The imported SSTables
+        // are now part of the live set, visible to reads.
+        class EachActivate extends AsyncFuture<Void> implements RequestCallbackWithFailure<Void>
+        {
+            final Set<InetAddressAndPort> responses = ConcurrentHashMap.newKeySet();
+
+            private EachActivate(Collection<InetAddressAndPort> peers)
+            {
+                responses.addAll(peers);
+            }
+
+            @Override
+            public void onResponse(Message<Void> msg)
+            {
+                logger.debug("Activation successfully applied on {}", msg.from());
+                SingleTransferResult result = CoordinatedTransfer.this.streams.get(msg.from());
+                if (result != null)
+                    result.completedActivation();
+
+                MutationTrackingService.instance.receivedActivationAck(CoordinatedTransfer.this, msg.from());
+                responses.remove(msg.from());
+                if (responses.isEmpty())
+                {
+                    // All activations complete, schedule cleanup to purge pending SSTables
+                    LocalTransfers.instance().scheduleCleanup();
+                    trySuccess(null);
+                }
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailure failure)
+            {
+                logger.error("Failed activation on {} due to {}", from, failure);
+                // TODO(expected): should only fail if we don't meet requested CL, individual failures are fine
+                responses.remove(from);
+                if (responses.isEmpty())
+                    trySuccess(null);
```

Comment:
Can we actually tolerate failures here? I could see this causing correctness problems if all requests fail because no replica received the promotion message and the caller believes that the transfer was successful. Or if a single node responded with a success response and then immediately went down. In both cases the caller believes the transfer succeeded, but it's actually unable to be completed.